Cf 2064-d (#6318)
cpinitiative/usaco-guide: solutions/silver/cf-2064D.mdx

---
id: cf-2064D
source: CF
title: Eating
author: Abhinav Singh
---
[Official Analysis (C++)](https://codeforces.com/blog/entry/138912)

## Explanation

Suppose the added slime has an initial weight of $x$. Since we process the slimes from right to left, define the suffix XOR array

$$
\text{suf}[i] = w_{i+1} \oplus w_{i+2} \oplus \cdots \oplus w_{n-1},
$$

> with $\text{suf}[n-1] = 0$. Then, after eating $k$ slimes, the slime's weight is simply $x \oplus \text{suf}[n-k-1].$

Reviewer: in my opinion, this isn't a very intuitive way of indexing suffix ? at least usually suf[i] would refer to the suffix including i

> Hence, when it attempts to eat the slime at position $i$, its current weight is $x \oplus \text{suf}[i]$. It succeeds if and only if $x \oplus \text{suf}[i] \ge w_i.$

Reviewer: imo no need to restate this

Let's consider what happens to our MSB when we eat that slime.
- If $\text{msb}(w_i) < \text{msb}(x \oplus \text{suf}[i])$, then, since XORing with a smaller value can't flip the leading bit, the MSB never decreases.
- If $\text{msb}(w_i) = \text{msb}(x \oplus \text{suf}[i])$, then the XOR cancels the leading bit, thus our MSB strictly decreases if we are able to eat the slime. Otherwise, we stop.

So we can greedily eat all the slimes from right to left whose MSB $\lt$ our current weight's MSB. Since XOR is associative, the XOR of this entire block can be obtained in O(1) using our precomputed suffix XOR values. Then, the next slime we hit has MSB $\ge$ the current weight's MSB. Either we fail there, or we eat it and the MSB strictly decreases. Since every weight is less than $2^{30}$, the MSB can only decrease at most 30 $(log(2^{30}))$ times.

Now, to find the first position to the left whose MSB is **at least** ours, we maintain a prefix array $last$. For each position (i) and bit position (j), define $last[i][j]$ as the rightmost position at or before $i$ whose slime has MSB at least $j$. We can compute it as

$$
\text{last}[i][j] =
\begin{cases}

> i & \text{if } msb(w_i) \ge j,\\

Reviewer: ```suggestion
i & \text{if } \text{msb}(w_i) \ge j,\\
```

\text{last}[i-1][j] & \text{otherwise}.
\end{cases}
$$

<Optional title="Offline Approach">

We can insert every query value into a
binary trie before sweeping from position $n-1$ down to $0$.
At each position $i$, repeatedly find the query $v$ minimizing $v \oplus \text{suf}[i]$. If that minimum XOR is $< w_i$, the query is blocked here, so we record answer $n - i - 1$ and remove $v$ from the trie. Repeat this until the minimum XOR is $\ge w_i$, then move left. Also $O((n + q) \log \mathcal{A})$ but requires offline processing and a binary trie.

> </Optional>

Reviewer: this is a bit of a cliffhanger, imo you should just add a sentence or two to summarize the algo as a whole

## Implementation

> **Time Complexity:** $\mathcal{O}((n + q) \log \mathcal{A})$, where $\mathcal{A}$

Reviewer: capittal letters for n and q

is the maximum value in the array $w$.

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);

	int test_num;
	cin >> test_num;

	for (int t = 0; t < test_num; t++) {
		int n, q;
		cin >> n >> q;

		vector<int> w(n + 1);
		for (int i = 1; i <= n; i++) { cin >> w[i]; }

		vector<int> suf(n + 1);

> 		for (int i = n - 1; i >= 0; i--) { suf[i] = suf[i + 1] ^ w[i + 1]; }

Reviewer: is 1 indexing w needed? we usually stick to 0 indexing if it makes sense

## Explanation

Define **suffix XOR** values $\text{suf}[i] = w_{i+1} \oplus \cdots \oplus w_{n-1}$ with $\text{suf}[n-1] = 0$. When a new slime of weight $x$ eats its way left, after consuming slimes at positions $n-1, n-2, \ldots, n-k$, its weight becomes $x \oplus \text{suf}[n-k-1]$.
So it can eat the slime at position $i$ iff $x \oplus \text{suf}[i] \geq w_i$, and the score is the number of slimes eaten before this first fails.

To answer all $q$ queries offline, we can load every query value into a binary trie. Then we sweep $i$ down from $n-1$ to $0$, maintaining $\text{curx} = \text{suf}[i]$ (the XOR of everything to the right of position $i$).
At each position $i$, query $v$ is blocked(i.e. the new slime with weight $v$ cannot eat the slime at position $i$) iff $v \oplus \text{curx} < w_i$. We repeatedly call `min_xor(curx)` on the trie - if the minimum XOR across all remaining queries is $< w_i$, that query is blocked here (score $= n - i - 1$), so we remove it and repeat.
If the minimum XOR is $\geq w_i$, then no remaining query is blocked at this position, so we can update $\text{curx} \mathrel{\oplus}= w_i$ and move to position $i - 1$.

> Whatever query is still in the trie after the full sweep has eaten all the $n$ slimes.

Reviewer 2: ```suggestion
Consider a query $v$ as "blocked to" $i$ if the added slime with weight $v$ stops eating the slime at position $i$ first.
At each position $i$, a query $v$ is blocked to $i$ if and only if $v \oplus \text{suf}[i] < w_i$.
We then use the trie to find the query with the minimum XOR value with $\text{suf}[i]$, and remove that query from the trie if the XOR value is $<w_i$, as that query cannot proceed past $i$; we do this repeatedly until the query with the minimum XOR in the trie *can* proceed past $i$, then move to position $i-1$.
```

## Implementation

> **Time Complexity:** $\mathcal{O}((n + q) \log \max w)$ - each query is inserted and removed from the trie at most once.

Reviewer 2: ```suggestion
**Time Complexity:** $\mathcal{O}((n + q) \log \max w)$, as each query is inserted and removed from the trie at most once.
```

---
id: cf-2064D
source: CF
title: Eating
author: Abhinav Singh
---

[Official Analysis (C++)](https://codeforces.com/blog/entry/138912)

## Explanation

Suppose the added slime has an initial weight of $x$.
Since XOR is associative and commutative, after eating $k$ slimes its weight is

$$
x \oplus w_{n-k} \oplus w_{n-k+1} \oplus \cdots \oplus w_{n-1}.
$$

Thus, we define **suffix XOR** values:
$$
\text{suf}[i] = w_{i+1} \oplus w_{i+2} \oplus \cdots \oplus w_{n-1},
$$
with $\text{suf}[n-1] = 0$. Then, after eating $k$ slimes, the slime's weight is simply $$x \oplus \text{suf}[n-k-1]$$.

Hence, when it attempts to eat the slime at position $i$, its
current weight is $x \oplus \text{suf}[i]$.
It succeeds if and only if $$x \oplus \text{suf}[i] \ge w_i.$$
Therefore, the answer to a query is the number of consecutive
slimes it can eat, sweeping from position $n-1$ down to $0$ before

> this inequality first fails. If it fails at position $i$, the

Reviewer 3: can you write the inequality in such a way that the reader don't have to look for it in previous sentences

slime has eaten positions $i + 1, \cdots n - 1$ for a score of $n- i - 1$.
If it never fails, the slime eats all $n$ slimes.

We process all queries offline, inserting every query value into a
binary trie before sweeping from position $n-1$ down to $0$.
Consider a query $v$ as "blocked at" $i$ if the added slime with
weight $v$ stops eating the slime at position $i$ first. At each
position $i$, a query $v$ is blocked at $i$ if and only if $v
\oplus \text{suf}[i] < w_i$. We can then use the trie to find the
query with the minimum XOR value with $\text{suf}[i]$, and remove
that query from the trie if the XOR value is $<w_i$, as that query
cannot proceed past $i$; we do this repeatedly until the query
with the minimum XOR in the trie *can* proceed past $i$, then move
to position $i-1$.

## Implementation

> **Time Complexity:** Since queries are removed permanently, each query is extracted at most once,

Reviewer 3: Just mention the total time complexity

---
id: cf-2064D
source: CF
title: Eating
author: Abhinav Singh
---

[Official Analysis(C++)](https://codeforces.com/blog/entry/138912)

## Explanation

> Define **suffix XOR** values $\text{suf}[i] = w_{i+1} \oplus \cdots \oplus w_{n-1}$ with $\text{suf}[n-1] = 0$. When a new slime of weight $x$ eats its way left, after consuming slimes at positions $n-1, n-2, \ldots, n-k$, its weight becomes $x \oplus \text{suf}[n-k-1]$.

Reviewer 3: these feels more code-oriented rather than an explanation. instead of justifying the code can you explain how can one arrive at this

Reviewer 2: @Sosuke23 I rewrote this a bit as the associativity of XOR should be obvious.  Can you check it out to see if this is good now?

	BinaryTrie() { trie.resize(1); }

	int new_node() {
		trie.push_back(Node());
		return (int)trie.size() - 1;
	}

	void add(int val) {
		int x = 0;

		for (int i = MAX_BIT; i >= 0; --i) {
			trie[x].cnt++;

			int b = (val >> i) & 1;

			if (trie[x].child[b] == -1) {
				trie[x].child[b] = new_node();
			} 

			x = trie[x].child[b];
		}

		trie[x].cnt++;
	}

	void remove(int val) {
		int x = 0;

		for (int i = MAX_BIT; i >= 0; --i) {
			trie[x].cnt--;

			int b = (val >> i) & 1;

			x = trie[x].child[b];
		}

		trie[x].cnt--;
	}

	int min_xor(int x) {
		int node = 0, res = 0;

		for (int i = MAX_BIT; i >= 0; --i) {
			int b = (x >> i) & 1;

			if (trie[node].child[b] != -1 && trie[trie[node].child[b]].cnt > 0) {
				node = trie[node].child[b];
			} else {
				res |= (1 << i);
				node = trie[node].child[b ^ 1];
			}
		}

		return res;
	}
};
// EndCodeSnip{}

int main() {

	int test_num;
	cin >> test_num;

	for (int t = 0; t < test_num; test_num++) {

		int n, q;
		cin >> n >> q;

		vector<long long> w(n);
		for (auto &x : w) cin >> x;

		vector<long long> queries(q);

		BinaryTrie trie;

		for (int i = 0; i < q; i++) {
			cin >> queries[i];
			trie.add((int)queries[i]);
		}

> 		map<long long, long long> answer;

Reviewer 3: if you are not taking care of order you can use unordered_map

Reviewer 2: I went ahead and pushed this since it was a relatively minor change.

---
id: cf-2064D
source: CF
title: Eating
author: Abhinav Singh
---

> [Official Analysis(C++)](https://codeforces.com/blog/entry/138912)

Reviewer 2: ```suggestion
[Official Analysis (C++)](https://codeforces.com/blog/entry/138912)
```

## Explanation

> Define **suffix XOR** values $\text{suf}[i] = w_{i+1} \oplus \cdots \oplus w_{n-1}$ with $\text{suf}[n-1] = 0$. When a new slime of weight $x$ eats its way left, after consuming slimes at positions $n-1, n-2, \ldots, n-k$, its weight becomes $x \oplus \text{suf}[n-k-1]$.

Reviewer 2: ```suggestion
Define **suffix XOR** values $\text{suf}[i] = w_{i+1} \oplus \cdots \oplus w_{n-1}$ with $\text{suf}[n-1] = 0$. When a new slime of weight $x$ eats $k$ slimes on its way to the left, its weight becomes $x \oplus \text{suf}[n-k-1]$.
```

$$(x \oplus w_{n-1}) \oplus w_{n-2} = x \oplus w_{n-1} \oplus w_{n-2},$$

> since XOR is associative. Continuing this process, after eating the last $k$ slimes, its weight is

Reviewer 2: Get rid of this paragraph—the derivation should be easy.  Just say “by XOR’s associativity…” and state the next equation

> So it can eat the slime at position $i$ iff $x \oplus \text{suf}[i] \geq w_i$, and the score is the number of slimes eaten before this first fails.

Reviewer 2: “So it can” is awkward also avoid iff jargon

> To answer all $q$ queries offline, we can load every query value into a binary trie. Then we sweep $i$ down from $n-1$ to $0$, maintaining $\text{curx} = \text{suf}[i]$ (the XOR of everything to the right of position $i$).

Reviewer 2: ```suggestion
To answer all $q$ queries offline, we load every query value into a binary trie. Then, we enumerate $i$ from $n-1$ to $0$ decreasing, maintaining $\text{suf}[i]$ dynamically, the XOR of everything to the right of position $i$.
```

Reviewer 2: Just use suf[i] throughout, makes it easier as cumulative postfix sum is an established optimization.
